### PR TITLE
[WIP][v7r1 -> v7r2] OAuth2/OIDC AuthN/AuthZ mechanism(depends on #4405)

### PR DESCRIPTION
**Pull request by TaykYoku**
This PR introduces the AuthN/AuthZ mechanism to DIRAC based on the use of Identity Provider
services using OAuth2/OIDC protocols.

Few remarks about the implementation. 

The user Authn/AuthZ paradigm changes. Before we collected all the necessary user information 
in the CS Registry and then used it to identify and authorize client requests. Now we can use
external identity providers dynamically. Each time a user is starting a DIRAC session, the 
authentication can be delegated to an external identity provider. If authentication is successful,    
then actual user profile information is cached in a dynamic session object which is created
for each active user. This session is used for further identification of user requests. Therefore,
DIRAC CS Registry becomes just one of the possible sources of the user profile information. Other
sources are (Federated) Identity Providers, e.g. Check-In, Indigo AIM, Google, etc, VOMS. 

The use of X509 certificates is no more the only way for user identification. Therefore,
certificate DN's are no more the primary user identifiers. Now there can be also user IDs 
from various identity providers. Within DIRAC, user name becomes the main, single and unique 
user identifier independent on various user identities from different providers. As a consequence, 
all the CS options with user identities as values, e.g. pilot user, should be expressed in terms 
of user names rather than DNs. For backward compatibility current options in a form of DN are 
still accepted but should be replaced eventually.  

There can be now DIRAC users that do not have usual personal certificates and will identified
via OAuth2/OIDC mechanism. However, in order to ensure the work of the DISET protocol still based
on the use of X509 certificate proxies, the us of Proxy Provider services is enabled. These
services can create a certificate proxy on demand to be used by users in CLI interfaces and
by DIRAC services for performing operation on the users' behalf. Current solution is DIRAC Proxy 
Provider using DIRAC CA certificates to generate user proxies (for the DIRAC internal use only). 
Solution using RCauth Proxy Provider service is tested and will be available in later PRs. 
Users having usual X509 certificates will continue to use those in a usual way.   

Identity Providers and Proxy Providers are added as new types of Resources. 

Although this PR introduces new functionalities, the current mechanisms of X509 based user
authentication are still maintained. Installations and users using old good X509 certificates
will continue to work without any changes.

BEGINRELEASENOTES

NEW: Multiple changes to provide OAuth2/OIDC user AuthN/AuthZ and OAuth session management.
CHANGE: multiple changes to use username instead of DN where applicable. 
                Changes affect the following Systems: Transformation, Configuration, DataManagement, Framework, RequestManagement, WorkloadManagement. Also tests, Core, Interfaces, Resources.

*ConfigurationSystem
CHANGE: Registry - modified for using cached users information on the clients

*Core
NEW: DictCache - add getDict method
CHANGE: AuthManager - split authorization logic, modify to use new Registry
CHANGE: RequestHandler - fill credDict by AuthManager methods
CHANGE: align with the PR changes

*FrameworkSystem
CHANGE: ProxyManager - modify to use user/group in requests, add VOMS information cache
NEW: dirac-proxy-init - able to use OAuth authentication flow
NEW: halo - new class to use spinners for waiting process, for ex. waiting authentication

*RequestManagementSystem
CHANGE: Request - add owner argument, use AuthManager methods
CHANGE: align with the PR changes

*WorkloadManagementSystem
CHANGE: add "no group" test, use Owner/username instead DN

*TransformationSystem
CHANGE: align with the PR changes

*DataManagementSystem
CHANGE: align with the PR changes

*Resources
CHANGE: align with the PR changes

*Interfaces
CHANGE: align with the PR changes

*tests
CHANGE: align with the PR changes

ENDRELEASENOTES

### CODE REVIEW

**Comment by chaen** [comment]
Unfortunately the changes done in the RMS are orthogonal to the development I have prepared.

File: Core/Base/API.py [hunk 1]
```diff
@@ -107,30 +119,45 @@ def _errorReport(self, error, message=None):
...
       return self._errorReport('Proxy information does not contain the group', res['Message'])
-    res = getDNForUsername(proxyInfo['username'])
-    if not res['OK']:
-      return self._errorReport('Failed to get proxies for user', res['Message'])
+    result = getDNsForUsername(proxyInfo['username'])
+    if not result['OK']:
+      return self._errorReport('Failed to get proxies for user', "No DNs found for %s: %s" % (proxyInfo['username'],
+                                                                                              result['Message']))
+    if not result['Value']:
+      return self._errorReport('Failed to get proxies for user', "No DNs found for %s" % proxyInfo['username'])
```
[the remark this answers is not in the document]
Comment (TaykYoku) [reply Core/Base/API.py]: corrected

File: ConfigurationSystem/Client/Helpers/Registry.py [hunk 2]
```diff
@@ -7,48 +7,73 @@
...
+  """ Return IDs for DIRAC user
+
+      :param str username: DIRAC user
+
+      :return: list -- contain IDs
+  """
+  return gConfig.getValue("%s/Users/%s/ID" % (gBaseRegistrySection, username), [])
+
+
+def getUsernameForID(ID, usersList=None):
```
[the remark this answers is not in the document]
Comment (TaykYoku) [reply ConfigurationSystem/Client/Helpers/Registry.py]: fixed

File: ConfigurationSystem/Client/Helpers/Registry.py [hunk 3]
```diff
@@ -7,48 +7,73 @@
...
+  return S_ERROR("No username found for ID %s" % ID)
 
 
-def getGroupsForDN(dn):
-  retVal = getUsernameForDN(dn)
-  if not retVal['OK']:
-    return retVal
-  return getGroupsForUser(retVal['Value'])
+def __getGroupsWithAttr(attrName, value):
+  """ Get all posible groups with some attribute
```
[the remark this answers is not in the document]
Comment (TaykYoku) [reply ConfigurationSystem/Client/Helpers/Registry.py]: corrected

File: ConfigurationSystem/Client/Helpers/Registry.py [hunk 4]
```diff
@@ -89,115 +141,84 @@ def getHostnameForDN(dn):
...
       name = result['Value']
     return getPropertiesForHost(name, defaultValue)
   else:
     return getPropertiesForGroup(group, defaultValue)
 
 
 def __matchProps(sProps, rProps):
+  """ Match properties
+
+      :param sProps: submited properties
```
[the remark this answers is not in the document]
Comment (TaykYoku) [reply ConfigurationSystem/Client/Helpers/Registry.py]: corrected

File: ConfigurationSystem/Client/Helpers/Registry.py [hunk 5]
```diff
@@ -361,147 +477,434 @@ def getVOMSRoleGroupMapping(vo=''):
...
-  for username in getUsersInGroup(groupName, defaultValue=[]):
-    email = gConfig.getValue("%s/Users/%s/Email" % (gBaseRegistrySection, username), [])
+  for username in getUsersInGroup(groupName):
+    email = getUserOption(username, 'Email', [])
     emails.append(email)
   return emails
+
+
+def getGroupsForDN(userDN, researchedGroup=None):
+  """ Get all posible groups for DN
```
[the remark this answers is not in the document]
Comment (TaykYoku) [reply ConfigurationSystem/Client/Helpers/Registry.py]: corrected

File: ConfigurationSystem/Client/Helpers/Registry.py [hunk 6]
```diff
@@ -361,147 +477,434 @@ def getVOMSRoleGroupMapping(vo=''):
...
+      :return: S_OK(list)/S_ERROR()
   """
-  result = getUsernameForDN(userDN)
+  try:
+    gSessionManager
+  except Exception:
+    try:
+      from OAuthDIRAC.FrameworkSystem.Client.OAuthManagerClient import gSessionManager  # pylint: disable=import-error
+    except Exception as ex:
+      return S_ERROR('Session manager not found:', ex)
```
[the remark this answers is not in the document]
Comment (TaykYoku) [reply ConfigurationSystem/Client/Helpers/Registry.py]: In order to support expected standard structure of result, for parsing as usual:
```
result = Registry.getProviderForID(self.__credDict['ID']) 
if not result['OK']:
   self.log.error(result['Message'])
   return False
```

**Comment by TaykYoku** [comment]
@chaen RMS main logic should not have changed except for the using user/group instead DN/group in some cases.

**Comment by chaen** [comment]
Yes I see, but my comment still holds

**Comment by TaykYoku** [comment]
@chaen can I see development that you have prepared for a better understanding of the issue

**Comment by chaen** [comment]
Nothing pushed yet, but I invite you to browse through the minutes of the BILD meeting where the changes were discussed and approved

**Comment by atsareg** [comment]
@chaen Can you please provide a link to the BiLD meeting where the RMS changes were discussed. This PR is certainly not affecting the RMS logic, neither the current one nor the future one

**Pull request closed without merging**